File: mvn-clean.py
import os
import subprocess
import sys
import shutil
from multiprocessing import Pool, Manager
from typing import List, Tuple

ERROR = "\u274C"  # unicode character for error messages
INFO = "\u2139"  # unicode character for informational messages

def get_dir_size(path: str) -> int:
    """
    Returns the total size in bytes of the directory at 'path', skipping symlinks and unreadable files.
    Faster and more robust using os.scandir.
    """
    total = 0
    try:
        with os.scandir(path) as it:
            for entry in it:
                try:
                    if entry.is_symlink():
                        continue
                    if entry.is_file(follow_symlinks=False):
                        total += entry.stat(follow_symlinks=False).st_size
                    elif entry.is_dir(follow_symlinks=False):
                        total += get_dir_size(entry.path)
                except Exception:
                    continue
    except Exception:
        pass
    return total

def clean_maven_in_target_dir(item_path: str, mvn_path: str) -> Tuple[str, str]:
    """
    Runs 'mvn clean' in the specified directory.
    Returns a tuple: (item_path, error_message). error_message is empty if successful.
    """
    try:
        subprocess.run([mvn_path, "-q", "clean"], cwd=item_path, check=True)
        print(f"Maven clean executed successfully in '{item_path}'.")
        return (item_path, "")
    except subprocess.CalledProcessError as e:
        return (item_path, f"Error executing mvn clean: {e}")
    except OSError as e:
        return (item_path, f"OS error: {e}")
def find_target_dirs(base_path: str) -> List[str]:
    """
    Finds Maven project directories (with 'pom.xml' and 'target' folder) in the base directory.
    """
    target_dirs = []
    for item in os.listdir(base_path):
        item_path = os.path.join(base_path, item)
        target_folder = os.path.join(item_path, "target")
        pom = os.path.join(item_path, "pom.xml")
        if os.path.exists(target_folder) and os.path.isdir(target_folder) and os.path.exists(pom):
            target_dirs.append(item_path)
    return target_dirs

def format_size(bytes_size: int) -> str:
    """
    Formats bytes as human-readable string.
    """
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_size < 1024:
            return f"{bytes_size:.2f} {unit}"
        bytes_size /= 1024
    return f"{bytes_size:.2f} PB"

def parse_args():
    import argparse
    parser = argparse.ArgumentParser(description="Clean Maven target directories in subfolders.")
    parser.add_argument("base_path", nargs="?", default=os.getcwd(), help="Base directory to search for Maven projects.")
    parser.add_argument("--dry-run", action="store_true", help="Show what would be cleaned and reclaimed, but do not run mvn clean.")
    return parser.parse_args()

def resolve_mvn_path(dry_run: bool) -> str:
    mvn_path = shutil.which("mvn.cmd") if os.name == "nt" else shutil.which("mvn")
    if mvn_path is None:
        mvn_path = shutil.which("mvn")
    if mvn_path is None and not dry_run:
        print(f"{ERROR} Maven (mvn) command not found in PATH. Please ensure Maven is installed and available.")
        sys.exit(1)
    return mvn_path

def main():
    """
    Main entry point for cleaning Maven target directories in subfolders.
    Supports optional --dry-run argument to only show what would be cleaned.
    """
    args = parse_args()
    base_path = args.base_path
    dry_run = args.dry_run


    if not os.path.exists(base_path):
        print(f"{ERROR} Base path '{base_path}' does not exist.")
        sys.exit(1)
    if not os.path.isdir(base_path):
        print(f"{ERROR} Base path '{base_path}' is not a directory.")
        sys.exit(1)

    mvn_path = resolve_mvn_path(dry_run)

    target_dirs = find_target_dirs(base_path)
    if not target_dirs:
        print(f"{ERROR} No Maven target directories found.")
        return

    print(f"{INFO} Found {len(target_dirs)} Maven project(s) to clean.")

    # Calculate total size of all target directories before cleaning
    total_bytes = 0
    target_sizes = {}
    for dir_path in target_dirs:
        target_folder = os.path.join(dir_path, "target")
        if os.path.exists(target_folder):
            size = get_dir_size(target_folder)
            target_sizes[dir_path] = size
            total_bytes += size

    print(f"{INFO} Total disk space that will be reclaimed: {format_size(total_bytes)}")

    print(f"\n{INFO} Directories to be cleaned:")
    for dir_path in target_dirs:
        size = target_sizes.get(dir_path, 0)
        print(f" - {dir_path} ({format_size(size)})")

    if dry_run:
        print(f"\n{INFO} Dry run: No directories were cleaned.")
        return

    # Multiprocessing-safe result collection
    with Pool(processes=os.cpu_count()) as pool:
        results = pool.starmap(clean_maven_in_target_dir, [(dir_path, mvn_path) for dir_path in target_dirs])

    cleaned = [path for path, err in results if not err]
    errors = [(path, err) for path, err in results if err]

    print(f"\n{INFO} Cleaned directories:")
    for c in cleaned:
        print(f" - {c}")
    print(f" {INFO} Total cleaned directories: {len(cleaned)}")

    if errors:
        print(f"\n{ERROR} Errors encountered:")
        for path, err in errors:
            print(f" - {path}: {err}")
        print(f" {ERROR} Total errors: {len(errors)}")

    # Only sum sizes for successfully cleaned directories
    reclaimed_bytes = sum(target_sizes.get(path, 0) for path in cleaned)
    print(f"\n{INFO} Total disk space reclaimed: {format_size(reclaimed_bytes)}")

if __name__ == "__main__":
    main()

File: mvn-clean.js
#!/usr/bin/env node

import { readdirSync, statSync, existsSync, rmSync } from 'node:fs';
import { resolve, join } from 'node:path';
import { spawn } from 'node:child_process';
import os from 'node:os';

const ERROR = '❌';
const INFO = 'ℹ️';

let abort = false;
let runningProcs = [];

process.on('SIGINT', () => {
    console.log('\nAborting...');
    abort = true;
    for (const proc of runningProcs) {
        proc.kill('SIGTERM');
    }
    runningProcs = [];
});

const spinnerChars = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
let spinnerIndex = 0;
let spinnerInterval;

function getDirSize(dirPath) {
    let total = 0;
    try {
        const entries = readdirSync(dirPath, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = join(dirPath, entry.name);
            try {
                if (entry.isSymbolicLink()) continue;
                if (entry.isFile()) {
                    total += statSync(fullPath).size;
                } else if (entry.isDirectory()) {
                    total += getDirSize(fullPath);
                }
            } catch { continue; }
        }
    } catch { }
    return total;
}

function formatSize(bytes) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let i = 0;
    while (bytes >= 1024 && i < units.length - 1) {
        bytes /= 1024;
        i++;
    }
    return `${bytes.toFixed(2)} ${units[i]}`;
}

function findProjectsToClean(basePath) {
    const projects = [];
    for (const item of readdirSync(basePath)) {
        const itemPath = join(basePath, item);
        if (!statSync(itemPath).isDirectory()) continue;
        const pom = join(itemPath, 'pom.xml');
        const packageJson = join(itemPath, 'package.json');
        const targetFolder = join(itemPath, 'target');
        const nodeModules = join(itemPath, 'node_modules');
        if (existsSync(pom) && existsSync(targetFolder)) {
            projects.push({ path: itemPath, type: 'maven', cleanFolder: 'target' });
        } else if (existsSync(packageJson) && existsSync(nodeModules)) {
            projects.push({ path: itemPath, type: 'node', cleanFolder: 'node_modules' });
        }
    }
    return projects;
}


function resolveMvnPath() {
    if (os.platform() === 'win32') {
        return 'mvn.cmd';
    }
    return 'mvn';
}

function parseArgs() {
    const args = process.argv.slice(2);
    let basePath = process.cwd();
    let dryRun = false;
    if (args.length > 0) {
        if (args.includes('--dry-run')) {
            dryRun = true;
            basePath = args.find(a => a !== '--dry-run') || basePath;
        } else {
            basePath = args[0];
        }
    }
    return { basePath, dryRun };
}

function validateBasePath(basePath) {
    if (!existsSync(basePath)) {
        console.error(`${ERROR} Base path '${basePath}' does not exist.`);
        process.exit(1);
    }
    if (!statSync(basePath).isDirectory()) {
        console.error(`${ERROR} Base path '${basePath}' is not a directory.`);
        process.exit(1);
    }
}

function reportProjects(projects, sizes) {
    console.log(`${INFO} Found ${projects.length} project(s) to clean.`);
    let totalBytes = Object.values(sizes).reduce((a, b) => a + b, 0);
    console.log(`${INFO} Total disk space that will be reclaimed: ${formatSize(totalBytes)}`);
    console.log(`\n${INFO} Projects to be cleaned:`);
    for (const proj of projects) {
        const size = sizes[proj.path] || 0;
        console.log(` - ${proj.path} (${proj.type}) (${formatSize(size)})`);
    }
}

async function cleanProjectAsync(proj) {
    const { path, type, cleanFolder } = proj;
    if (type === 'maven') {
        const mvnPath = resolveMvnPath();
        return new Promise((resolve) => {
            const proc = spawn(`${mvnPath} -q clean`, [], { cwd: path, shell: true });
            runningProcs.push(proc);
            proc.on('close', (code) => {
                runningProcs = runningProcs.filter(p => p !== proc);
                if (code === 0) {
                    resolve({ path, error: '' });
                } else {
                    resolve({ path, error: `Error executing mvn clean: exit code ${code}` });
                }
            });
            proc.on('error', (e) => {
                runningProcs = runningProcs.filter(p => p !== proc);
                resolve({ path, error: `OS error: ${e.message}` });
            });
        });
    } else if (type === 'node') {
        try {
            rmSync(join(path, cleanFolder), { recursive: true, force: true });
            return { path, error: '' };
        } catch (e) {
            return { path, error: `Error removing ${cleanFolder}: ${e.message}` };
        }
    } else {
        return { path, error: 'Unknown project type' };
    }
}

async function cleanProjects(projects, concurrency = os.cpus().length) {
    const cleaned = [];
    const errors = [];
    let shared = { index: 0, processed: 0 };

    spinnerInterval = setInterval(() => {
        process.stdout.write(`\r${spinnerChars[spinnerIndex++ % spinnerChars.length]} Cleaning... ${shared.processed}/${projects.length}`);
    }, 100);

    async function worker() {
        while (true) {
            if (abort) break;
            let proj;
            if (shared.index < projects.length) {
                proj = projects[shared.index++];
            } else {
                break;
            }
            const result = await cleanProjectAsync(proj);
            shared.processed++;
            if (result.error) {
                errors.push(result);
            } else {
                cleaned.push(result.path);
            }
        }
    }

    await Promise.all(new Array(concurrency).fill().map(worker));
    clearInterval(spinnerInterval);
    if (abort) {
        process.stdout.write('\rAborted.\n');
    } else {
        process.stdout.write('\rDone cleaning.\n');
    }
    return { cleaned, errors };
}

function reportCleaned(cleaned, errors, sizes) {
    console.log(`\n${INFO} Cleaned projects:`);
    for (const c of cleaned) {
        console.log(` - ${c}`);
    }
    console.log(` ${INFO} Total cleaned projects: ${cleaned.length}`);

    if (errors.length > 0) {
        console.log(`\n${ERROR} Errors encountered:`);
        for (const { path, error } of errors) {
            console.log(` - ${path}: ${error}`);
        }
        console.log(` ${ERROR} Total errors: ${errors.length}`);
    }

    const reclaimedBytes = cleaned.reduce((sum, path) => sum + (sizes[path] || 0), 0);
    console.log(`\n${INFO} Total disk space reclaimed: ${formatSize(reclaimedBytes)}`);
}

function getCleanSizes(projects) {
    const sizes = {};
    for (const proj of projects) {
        const folder = join(proj.path, proj.cleanFolder);
        if (existsSync(folder)) {
            sizes[proj.path] = getDirSize(folder);
        }
    }
    return sizes;
}

async function main() {
    const { basePath, dryRun } = parseArgs();
    validateBasePath(basePath);

    const startTime = process.hrtime.bigint();

    const projects = findProjectsToClean(basePath);

    if (projects.length === 0) {
        const absolutePath = resolve(basePath);
        console.log(`${ERROR} No projects found to clean in '${absolutePath}'.`);
        return;
    }

    const sizes = getCleanSizes(projects);
    reportProjects(projects, sizes);

    if (dryRun) {
        console.log(`\n${INFO} Dry run: No projects were cleaned.`);
        return;
    }

    const { cleaned, errors } = await cleanProjects(projects);
    reportCleaned(cleaned, errors, sizes);

    const endTime = process.hrtime.bigint();
    const timeTaken = Number(endTime - startTime) / 1e9;
    console.log(`\n${INFO} Time taken: ${timeTaken.toFixed(2)}s`);
}

await main();
